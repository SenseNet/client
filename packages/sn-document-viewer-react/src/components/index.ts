export * from './comment'
export * from './document-widgets'
export * from './page-widgets'
export * from './ConfirmationDialog'
export * from './DocumentViewer'
export * from './DocumentViewerError'
export * from './DocumentViewerLayout'
export * from './DocumentViewerLoading'
export * from './DocumentViewerRegeneratePreviews'
export * from './LayoutAppBar'
export * from './Page'
export * from './PageList'
export * from './styles'
