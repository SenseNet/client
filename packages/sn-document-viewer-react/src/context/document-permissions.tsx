import React, { useEffect, useState } from 'react'
import { useDocumentData, useDocumentViewerApi } from '../hooks'

export const DocumentPermissionsContext = React.createContext({
  canEdit: false,
  canHideRedaction: false,
  canHideWatermark: false,
})

export const DocumentPermissionsContextProvider: React.FC = ({ children }) => {
  const { documentData } = useDocumentData()
  const api = useDocumentViewerApi()
  const [canEdit, setcanEdit] = useState(false)
  const [canHideRedaction, setcancanHideRedaction] = useState(false)
  const [canHideWatermark, setcanHideWatermark] = useState(false)

  useEffect(() => {
    const abortController = new AbortController()
    ;(async () => {
      try {
        const canEditPromise = api.canEditDocument({ document: documentData, abortController })
        const cancanHideRedactionPromise = api.canHideRedaction({ document: documentData, abortController })
        const canHideWatermarkPromise = api.canHideWatermark({ document: documentData, abortController })

        const [canEditValue, cancanHideRedactionValue, canHideWatermarkValue] = await Promise.all([
          canEditPromise,
          cancanHideRedactionPromise,
          canHideWatermarkPromise,
        ])
        setcanEdit(canEditValue)
        setcancanHideRedaction(cancanHideRedactionValue)
        setcanHideWatermark(canHideWatermarkValue)
      } catch (error) {
        if (!abortController.signal.aborted) {
          throw error
        }
      }
    })()
  }, [api, documentData])

  return (
    <DocumentPermissionsContext.Provider
      value={{
        canEdit,
        canHideRedaction,
        canHideWatermark,
      }}>
      {children}
    </DocumentPermissionsContext.Provider>
  )
}
