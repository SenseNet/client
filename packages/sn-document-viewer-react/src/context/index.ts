export * from './api-settings'
export * from './comment-states'
export * from './comments'
export * from './document-data'
export * from './document-permissions'
export * from './localization-context'
export * from './preview-image-data'
export * from './viewer-settings'
export * from './viewer-state'
