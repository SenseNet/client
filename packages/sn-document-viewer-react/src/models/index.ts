export * from './DocumentViewerApiSettings'
export * from './Theming'
export * from './viewer-state'
