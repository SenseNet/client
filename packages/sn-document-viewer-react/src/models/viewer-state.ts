import { ObservableValue } from '@sensenet/client-utils'

/**
 * Rotation model
 */
export interface RotationModel {
  /**
   * The index of the page
   */
  pageNum: number

  /**
   * Rotation in degrees
   */
  degree: number
}

/**
 * Type model for the Viewer state
 */
export interface ViewerState {
  /**
   * The active page
   */
  activePage: number
  /**
   * Watermark is on / off
   */
  showWatermark: boolean
  /**
   * Redaction is on / off
   */
  showRedaction: boolean
  /**
   * Shapes are on / off
   */
  showShapes: boolean

  /**
   * Thumbnails on / off
   */
  showThumbnails: boolean

  /**
   * Zoom level relative to the fitted image size
   */
  zoomLevel: number

  /**
   * Rotation
   */
  rotation?: RotationModel[]

  /**
   * Determines if comments are shown
   */
  showComments: boolean

  /**
   * True if there are any changes on the document
   */
  hasChanges: boolean

  /**
   * Variable that indicates if a comment marker placing is in progress
   */
  isPlacingCommentMarker: boolean

  /**
   * Indicates if the user is creating a comment at the moment
   */
  isCreateCommentActive: boolean

  /**
   *
   */
  onPageChange: ObservableValue<number>
}
