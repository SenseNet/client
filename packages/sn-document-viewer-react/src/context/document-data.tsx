import { DocumentData } from '@sensenet/client-core'
import { deepMerge, DeepPartial, sleepAsync } from '@sensenet/client-utils'
import { useRepository } from '@sensenet/hooks-react'
import React, { useCallback, useEffect, useState } from 'react'
import Semaphore from 'semaphore-async-await'
import { POLLING_INTERVAL } from '../components'
import { PreviewState } from '../enums'
import { useDocumentViewerApi, useViewerSettings } from '../hooks'

const defaultDocumentData: DocumentData = {
  documentName: '',
  documentType: '',
  fileSizekB: 0,
  hostName: '',
  idOrPath: 0,
  pageCount: PreviewState.Loading,
  shapes: {
    annotations: [],
    highlights: [],
    redactions: [],
  },
  error: undefined,
}

export interface DocumentDataContextType {
  documentData: DocumentData
  updateDocumentData: (data: DeepPartial<DocumentData>) => void
  isInProgress: boolean
  triggerReload: () => void
}

export const defaultDocumentDataContextValue: DocumentDataContextType = {
  documentData: defaultDocumentData,
  updateDocumentData: () => undefined,
  isInProgress: false,
  triggerReload: () => {},
}

export const DocumentDataContext = React.createContext<DocumentDataContextType>(defaultDocumentDataContextValue)

export const DocumentDataProvider: React.FC = ({ children }) => {
  const api = useDocumentViewerApi()
  const doc = useViewerSettings()
  const repo = useRepository()

  const [documentData, setDocumentData] = useState<DocumentData>(defaultDocumentData)
  const [loadLock] = useState(new Semaphore(1))
  const [isInProgress, setIsInProgress] = useState(false)
  const [reloadToken, setReloadToken] = useState(0)

  const triggerReload = () => setReloadToken((prevValue) => prevValue + 1)

  useEffect(() => {
    const ac = new AbortController()
    const getData = async () => {
      const result = await api.getDocumentData({
        hostName: repo.configuration.repositoryUrl,
        idOrPath: doc.documentIdOrPath,
        version: doc.version,
        abortController: ac,
      })
      if (result.pageCount === PreviewState.Loading && ac.signal.aborted === false) {
        setDocumentData(result)
        await sleepAsync(POLLING_INTERVAL)
        getData()
      } else {
        setDocumentData(result)
      }
    }
    ;(async () => {
      try {
        setIsInProgress(true)
        await loadLock.acquire()
        await getData()
      } catch (error) {
        if (!ac.signal.aborted) {
          setDocumentData({ ...defaultDocumentData, pageCount: PreviewState.ClientFailure, error: error.toString() })
          throw error
        }
      } finally {
        setIsInProgress(false)
        loadLock.release()
      }
    })()
    return () => ac.abort()
  }, [api, doc.documentIdOrPath, doc.version, loadLock, repo.configuration.repositoryUrl, reloadToken])

  const updateDocumentData = useCallback(
    (newDocData: Partial<DocumentData>) => {
      const merged = deepMerge(documentData, newDocData)
      if (JSON.stringify(documentData) !== JSON.stringify(merged)) {
        setDocumentData(merged)
      }
    },
    [documentData],
  )

  if (!documentData.idOrPath) {
    return null
  }

  return (
    <DocumentDataContext.Provider value={{ documentData, updateDocumentData, isInProgress, triggerReload }}>
      {children}
    </DocumentDataContext.Provider>
  )
}
