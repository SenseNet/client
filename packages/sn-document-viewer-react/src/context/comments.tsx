import React, { useEffect, useState } from 'react'
import { Trace } from '@sensenet/client-utils'
import { useDocumentData, useDocumentViewerApi, useViewerState } from '../hooks'
import { CommentData } from '../models/Comment'
import { DocumentViewerApiSettings } from '../models'

export type CommentsContextType = {
  comments: CommentData[]
} & DocumentViewerApiSettings['commentActions']

export const defaultCommentsContext: CommentsContextType = {
  comments: [],
  addPreviewComment: async () => undefined as any,
  deletePreviewComment: async () => undefined as any,
  getPreviewComments: async () => undefined as any,
}

export const CommentsContext = React.createContext<CommentsContextType>(defaultCommentsContext)

export interface CommentContextProviderProps {
  page: number
  images: 'preview' | 'thumbnail'
}

export const CommentsContextProvider: React.FC<CommentContextProviderProps> = ({ page, children, images }) => {
  const api = useDocumentViewerApi()
  const { documentData } = useDocumentData()
  const viewerState = useViewerState()
  const [comments, setComments] = useState<CommentData[]>([])
  useEffect(() => {
    const abortController = new AbortController()
    ;(async () => {
      if (viewerState.showComments && images === 'preview') {
        const result = await api.commentActions.getPreviewComments({ document: documentData, page, abortController })
        setComments(result)
      }
    })()
    return () => abortController.abort()
  }, [api.commentActions, documentData, images, page, viewerState.showComments])

  useEffect(() => {
    const disposables = [
      Trace.method({
        object: api.commentActions,
        isAsync: true,
        method: api.commentActions.addPreviewComment,
        onFinished: async ({ returned }) => {
          const a = await returned
          setComments([...comments, a])
          viewerState.updateState({ isPlacingCommentMarker: false })
        },
      }),
      Trace.method({
        object: api.commentActions,
        method: api.commentActions.deletePreviewComment,
        isAsync: true,
        onFinished: async ({ returned, methodArguments }) => {
          const returnedValue = await returned
          if (returnedValue.modified) {
            setComments(comments.filter(c => c.id !== methodArguments[0].commentId))
          }
        },
      }),
    ]

    return () => disposables.forEach(d => d.dispose())
  }, [api.commentActions, comments, viewerState])

  return (
    <CommentsContext.Provider
      value={{
        ...api.commentActions,
        comments,
      }}>
      {children}
    </CommentsContext.Provider>
  )
}
