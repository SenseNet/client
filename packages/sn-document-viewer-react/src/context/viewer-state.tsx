import { deepMerge, DeepPartial, ObservableValue } from '@sensenet/client-utils'
import React, { useCallback, useEffect, useState } from 'react'
import { ViewerState } from '../models/viewer-state'

export const defaultViewerState: ViewerState & {
  updateState: (newState: DeepPartial<ViewerState>) => void
} = {
  activePage: 1,
  zoomMode: 'custom',
  showWatermark: false,
  showRedaction: true,
  showShapes: true,
  showThumbnails: false,
  fitRelativeZoomLevel: 0,
  rotation: undefined,
  showComments: false,
  hasChanges: false,
  isPlacingCommentMarker: false,
  isCreateCommentActive: false,
  onPageChange: new ObservableValue(1),
  updateState: () => {},
}
export const ViewerStateContext = React.createContext(defaultViewerState)

export const ViewerStateProvider: React.FC<{ options?: Partial<typeof defaultViewerState> }> = (props) => {
  const [state, setState] = useState<typeof defaultViewerState>(deepMerge({ ...defaultViewerState }, props.options))

  useEffect(() => {
    setState(deepMerge({ ...defaultViewerState }, props.options))
  }, [props.options])

  const updateState = useCallback(
    (newState: DeepPartial<typeof defaultViewerState>) => {
      setState(deepMerge({ ...state }, newState))
    },
    [state],
  )

  return <ViewerStateContext.Provider value={{ ...state, updateState }}>{props.children}</ViewerStateContext.Provider>
}
