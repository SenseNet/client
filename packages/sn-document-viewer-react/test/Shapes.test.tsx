import { mount, shallow } from 'enzyme'
import React from 'react'
import { ThemeProvider } from 'styled-components'
import {
  CommentMarker,
  ShapeAnnotation,
  ShapeHighlight,
  ShapeRedaction,
  ShapeSkeleton,
  ShapesWidget,
  ShapesWidgetProps,
} from '../src'
import { CommentsContext } from '../src/context/comments'
import { DocumentDataContext } from '../src/context/document-data'
import { DocumentPermissionsContext } from '../src/context/document-permissions'
import { defaultViewerState, ViewerStateContext } from '../src/context/viewer-state'
import { defaultTheme } from '../src/models'
import { exampleDocumentData, examplePreviewImageData } from './__Mocks__/viewercontext'

describe('Shapes component', () => {
  const defaultProps: ShapesWidgetProps = {
    page: examplePreviewImageData,
    zoomRatio: 1,
  }

  it('should render all the shapes', () => {
    const wrapper = mount(
      <DocumentDataContext.Provider value={{ documentData: exampleDocumentData } as any}>
        <DocumentPermissionsContext.Provider value={{ canEdit: true, canHideRedaction: true, canHideWatermark: true }}>
          <ViewerStateContext.Provider value={{ ...defaultViewerState, showRedaction: true, showShapes: true }}>
            <ShapesWidget {...defaultProps} />
          </ViewerStateContext.Provider>
        </DocumentPermissionsContext.Provider>
      </DocumentDataContext.Provider>,
    )
    expect(wrapper.find(ShapeRedaction).exists()).toBeTruthy()
    expect(wrapper.find(ShapeAnnotation).exists()).toBeTruthy()
    expect(wrapper.find(ShapeHighlight).exists()).toBeTruthy()
  })

  it('should render just redactions', () => {
    const wrapper = mount(
      <DocumentDataContext.Provider value={{ documentData: exampleDocumentData } as any}>
        <DocumentPermissionsContext.Provider value={{ canEdit: true, canHideRedaction: true, canHideWatermark: true }}>
          <ViewerStateContext.Provider value={{ ...defaultViewerState, showRedaction: true, showShapes: false }}>
            <ShapesWidget {...defaultProps} />
          </ViewerStateContext.Provider>
        </DocumentPermissionsContext.Provider>
      </DocumentDataContext.Provider>,
    )
    expect(wrapper.find(ShapeRedaction).exists()).toBeTruthy()
    expect(wrapper.find(ShapeAnnotation).exists()).toBeFalsy()
    expect(wrapper.find(ShapeHighlight).exists()).toBeFalsy()
  })

  it('should not render any shape or redactions', () => {
    const wrapper = shallow(<ShapesWidget {...defaultProps} />)
    expect(wrapper.find(ShapeRedaction).exists()).toBeFalsy()
    expect(wrapper.find(ShapeAnnotation).exists()).toBeFalsy()
    expect(wrapper.find(ShapeHighlight).exists()).toBeFalsy()
  })

  it('should render comment markers', () => {
    const wrapper = mount(
      <ThemeProvider theme={defaultTheme}>
        <ViewerStateContext.Provider value={{ ...defaultViewerState, showComments: true }}>
          <CommentsContext.Provider
            value={{
              comments: [
                { x: '10', y: '10', id: 'a', page: 1, text: 'a', createdBy: {} as any },
                { x: '20', y: '20', id: 'b', page: 1, text: 'b', createdBy: {} as any },
              ],
              addPreviewComment: jest.fn(),
              deletePreviewComment: jest.fn(),
              getPreviewComments: jest.fn(),
            }}>
            <ShapesWidget {...defaultProps} />
          </CommentsContext.Provider>
        </ViewerStateContext.Provider>
      </ThemeProvider>,
    )
    expect(wrapper.find(CommentMarker).exists()).toBeTruthy()
    expect(wrapper.find(CommentMarker).length).toBe(2)
  })

  it('should handle keyup/mouseUp events', () => {
    const updateState = jest.fn()
    const updateDocumentData = jest.fn()

    const wrapper = mount(
      <DocumentDataContext.Provider
        value={{ documentData: exampleDocumentData, updateDocumentData, isInProgress: false }}>
        <DocumentPermissionsContext.Provider value={{ canEdit: true, canHideRedaction: true, canHideWatermark: true }}>
          <ViewerStateContext.Provider
            value={{ ...defaultViewerState, showRedaction: true, showShapes: true, updateState }}>
            <ShapesWidget {...defaultProps} />
          </ViewerStateContext.Provider>
        </DocumentPermissionsContext.Provider>
      </DocumentDataContext.Provider>,
    )
    wrapper.find(ShapeSkeleton).last().simulate('keyUp', { key: 'Delete' })
    expect(updateDocumentData).toBeCalled()
    expect(updateState).toBeCalledWith({ hasChanges: true })
  })

  it('should handle onBlur event', () => {
    const updateState = jest.fn()
    const updateDocumentData = jest.fn()

    const wrapper = mount(
      <DocumentDataContext.Provider
        value={{ documentData: exampleDocumentData, updateDocumentData, isInProgress: false }}>
        <DocumentPermissionsContext.Provider value={{ canEdit: true, canHideRedaction: true, canHideWatermark: true }}>
          <ViewerStateContext.Provider
            value={{ ...defaultViewerState, showRedaction: true, showShapes: true, updateState }}>
            <ShapesWidget {...defaultProps} />
          </ViewerStateContext.Provider>
        </DocumentPermissionsContext.Provider>
      </DocumentDataContext.Provider>,
    )

    wrapper.find('#annotation-input').simulate('focus')
    wrapper.find('#annotation-input').simulate('blur')
    expect(updateDocumentData).toBeCalled()
    expect(updateState).toBeCalledWith({ hasChanges: true })
  })

  it('should handle drag & drop event', () => {
    const updateState = jest.fn()
    const updateDocumentData = jest.fn()

    const wrapper = mount(
      <DocumentDataContext.Provider
        value={{ documentData: exampleDocumentData, updateDocumentData, isInProgress: false }}>
        <DocumentPermissionsContext.Provider value={{ canEdit: true, canHideRedaction: true, canHideWatermark: true }}>
          <ViewerStateContext.Provider
            value={{ ...defaultViewerState, showRedaction: true, showShapes: true, updateState }}>
            <ShapesWidget {...defaultProps} />
          </ViewerStateContext.Provider>
        </DocumentPermissionsContext.Provider>
      </DocumentDataContext.Provider>,
    )

    wrapper
      .find(ShapeSkeleton)
      .last()
      .simulate('drop', {
        dataTransfer: {
          getData: () => {
            return JSON.stringify({
              type: 'highlights',
              shape: {
                guid: '9a324f30-1423-11e9-bcb9-d719ddfb5f43',
                imageIndex: 1,
                h: 100,
                w: 100,
                x: 100,
                y: 100,
              },
              offset: {
                width: 10,
                height: 10,
              },
            })
          },
        },
      })

    expect(updateDocumentData).toBeCalled()
    expect(updateState).toBeCalledWith({ hasChanges: true })
  })
})
