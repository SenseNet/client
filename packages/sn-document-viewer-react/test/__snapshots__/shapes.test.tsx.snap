// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shapes component should render ShapeDraft without crashing 1`] = `
<div
  className="makeStyles-root-33 makeStyles-root-34"
  key="draft-30-40"
  tabIndex={0}
/>
`;
